### PR TITLE
Added yaml file that include bias correction for atms_npp.

### PR DESCRIPTION
**Pull request by xyzemc**
Addressing [issue#52](https://github.com/NOAA-EMC/RDASApp/issues/52 ), this pull request include a new yaml file includes the bias correction part on top of rrfs_mpasjedi_2022052619_atms_npp_qc.yaml. This new yaml file named as 'rrfs_mpasjedi_2022052619_atms_npp_qc_**bc**.yaml'
Also, three fixed files have been added into the test data Data/obs directory: 
- atms_npp.satbias.nc4
- atms_npp.satbias_cov.nc4
- atms_npp.tlapse.txt

There will be a new output directory Data/bc, which contains two output bias files 
- out_atms_npp.satbias.nc  
- out_atms_npp.satbias_cov.nc

The ctest has been successfully passed, the test result as bellow:

> [Xiaoyan.Zhang@hfe03 rrfs-test]$ ctest
Test project /scratch2/NCEPDEV/fv3-cam/Xiaoyan.Zhang/noscrub/JEDI/RDASApp_atms/build/rrfs-test
    Start 1: rrfs_mpasjedi_2022052619_Ens3Dvar
1/4 Test #1: rrfs_mpasjedi_2022052619_Ens3Dvar .........   Passed   90.76 sec
    Start 2: rrfs_mpasjedi_2022052619_letkf
2/4 Test #2: rrfs_mpasjedi_2022052619_letkf ............   Passed   51.20 sec
    Start 3: rrfs_mpasjedi_2022052619_atms_npp_qc
3/4 Test #3: rrfs_mpasjedi_2022052619_atms_npp_qc ......   Passed   90.54 sec
    Start 4: rrfs_mpasjedi_2022052619_atms_npp_qc_bc
4/4 Test #4: rrfs_mpasjedi_2022052619_atms_npp_qc_bc ...   Passed  155.86 sec
100% tests passed, 0 tests failed out of 4
Label Time Summary:
mpi            = 388.37 sec*proc (4 tests)
rdas-bundle    = 388.37 sec*proc (4 tests)
script         = 388.37 sec*proc (4 tests)
Total Test time (real) = 388.72 sec

### CODE REVIEW

**Comment by SamuelDegelia-NOAA** [comment]
Hi @xyzemc, I am not seeing the yaml file in this PR yet.

**Comment by xyzemc** [comment]
> Hi @xyzemc, I am not seeing the yaml file or observations in this PR yet.

Thanks for checking. It should be there now.

File: modulefiles/RDAS/hera.intel.lua
Comment (TingLei-NOAA) [inline modulefiles/RDAS/hera.intel.lua]: why do we keep this commented out line here?  For easier switch back in the future?

File: rrfs-test/CMakeLists.txt
Comment (TingLei-NOAA) [inline rrfs-test/CMakeLists.txt]: We need a more compact way to deal with this "executable assigning" in the future.

File: modulefiles/RDAS/hera.intel.lua
Comment (xyzemc) [reply modulefiles/RDAS/hera.intel.lua]: > why do we keep this commented out line here? For easier switch back in the future?

Yes. The test data is in my directory now. But later, Sam will move them together to his directory. But I can remove the commented out line of your directory if you do not need to keep it.

File: rrfs-test/CMakeLists.txt
Comment (TingLei-NOAA) [reply rrfs-test/CMakeLists.txt]: That is fine!

**Review by TingLei-NOAA — DISMISSED** [review]

**Comment by SamuelDegelia-NOAA** [comment]
I copied the new atms fixed files for bias correction onto my staged data on Hera (`/scratch1/NCEPDEV/da/Samuel.Degelia/RRFS/rrfs-test-data/rrfs-data_mpasjedi_2022052619/Data/obs`). I will put the data on Orion whenever that maintenance ends. @guoqing-noaa will need to copy the new files into the corresponding directories on Jet. Hopefully we will have a role account soon that will make this process a little easier.

**Review by SamuelDegelia-NOAA — DISMISSED** [review]

**Comment by ShunLiu-NOAA** [comment]
@guoqing-noaa Do you get a chance to update the file on JET?

**Comment by guoqing-noaa** [comment]
> @guoqing-noaa Do you get a chance to update the file on JET?

`/scratch1/NCEPDEV/da/Samuel.Degelia/RRFS/rrfs-test-data/rrfs-data_mpasjedi_2022052619/Data/obs
`has been copied to Jet. Thanks!

**Review by TingLei-NOAA — DISMISSED** [review]

**Comment by ShunLiu-NOAA** [comment]
@xyzemc Could you please test your modification with the new MPAS test cases?

**Comment by xyzemc** [comment]
> @xyzemc Could you please test your modification with the new MPAS test cases?

will do

**Comment by xyzemc** [comment]
> @xyzemc Could you please test your modification with the new MPAS test cases?

I tried test it with new MPAS 12-km test case on 2024052700, but there is no atms_npp data cover the domain at this time. So the test can't be finished.

**Comment by guoqing-noaa** [comment]
@xyzemc will the Atlantic bigger domain have some atms_npp data?

**Comment by xyzemc** [comment]
> @xyzemc will the Atlantic bigger domain have some atms_npp data?

For the case on 2024052700, the RAP ATMS bufr data only have one swath that is over the pacific ocean. I can use global gdas bufr data to test your case. I will update the result here once I get it done.
The **globa**l atms_npp data at 2024052700 coverage is here:
image
But the **RAP** atms_npp data at 2024052700 coverage is much less:
image

**Comment by xyzemc** [comment]
Update for testing ATMS_NPP data in new 12-km domain:
**ATMS_NPP data source:** 2024052700.gdas.t00z.atms.tm00.bufr_d
**Converted to IODA data** : 2024052700.gdas.t00z.atms.npp.tm00.nc
**The radius used to do LAMDomainCheck is 2500km**

       **Obs passed QC**
image
  
         **HoFx**
image

            **O-B with QC**
image

             **Analysis Increment**
image

**Comment by xyzemc** [comment]
Ctest result:

> ctest -R rrfs_mpasjedi_2024052700_atms_npp_qc_bc
Test project /scratch2/NCEPDEV/fv3-cam/Xiaoyan.Zhang/noscrub/JEDI/RDASApp_08072024/build/rrfs-test
    Start 6: rrfs_mpasjedi_2024052700_atms_npp_qc_bc
1/1 Test #6: rrfs_mpasjedi_2024052700_atms_npp_qc_bc ...   Passed  313.06 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
mpi            = 313.06 sec*proc (1 test)
rdas-bundle    = 313.06 sec*proc (1 test)
script         = 313.06 sec*proc (1 test)

Total Test time (real) = 314.06 sec

**Review by TingLei-NOAA — DISMISSED** [review]

**Comment by SamuelDegelia-NOAA** [comment]
As discussed today @xyzemc, could you update this PR to just include the new `rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml`? We do not plan to add a ctest specifically for each observation type, so the rest of the changes are not needed at the moment. Thank you!

**Comment by SamuelDegelia-NOAA** [comment]
Hi @xyzemc thank you for making the updates! I am trying to run this for the MPAS test case by inserting the obs space section from `atms_npp_qc_bc.yaml` into `rrfs_mpasjedi_2024052700_Ens3Dvar.yaml`. But I get the following error: 

```
 20: Exception: lapse_rate does not exist in ufo::PredictorFactory
 20: Exception: oops::Variational<MPAS, UFO and IODA observations> terminating...
 20: Exception: level: 0
```

Also, I have a couple of suggested changes. Could you do the following: 

1. Add the following to the final line of `RDASApp/rrfs-test/scripts/link_mpasjedi_expr.sh` so that we can link the CRTM coefficients into the new `expr/` directory: 

     `ln -snf ${RDASApp}/fix/crtm/2.4.0 crtm_2.4.0`

2. The `rrfs_mpasjedi_2022052619_atms_npp_qc_bc.yaml` yaml added in this PR is for the old MPAS-JEDI case that has been deprecated in RDASApp. Could you either (1) update this yaml for the new test case, or (2) remove this file from the PR since the addition to `validated_yamls` covers the same material.

**Comment by xyzemc** [comment]
> Hi @xyzemc thank you for making the updates! I am trying to run this for the MPAS test case by inserting the obs space section from `atms_npp_qc_bc.yaml` into `rrfs_mpasjedi_2024052700_Ens3Dvar.yaml`. But I get the following error:
> 
> ```
>  20: Exception: lapse_rate does not exist in ufo::PredictorFactory
>  20: Exception: oops::Variational<MPAS, UFO and IODA observations> terminating...
>  20: Exception: level: 0
> ```
> 
> Also, I have a couple of suggested changes. Could you do the following:
> 
> 1. Add the following to the final line of `RDASApp/rrfs-test/scripts/link_mpasjedi_expr.sh` so that we can link the CRTM coefficients into the new `expr/` directory:
>    `ln -snf ${RDASApp}/fix/crtm/2.4.0 crtm_2.4.0`
> 2. The `rrfs_mpasjedi_2022052619_atms_npp_qc_bc.yaml` yaml added in this PR is for the old MPAS-JEDI case that has been deprecated in RDASApp. Could you either (1) update this yaml for the new test case, or (2) remove this file from the PR since the addition to `validated_yamls` covers the same material.

For the error message, please check if the 'atms_npp.tlapse.txt' has been linked/copied successfully.

I have add ' `ln -snf ${RDASApp}/fix/crtm/2.4.0 crtm_2.4.0` and removed the old case yaml file, and committed it. Please double check. Thanks!

**Comment by SamuelDegelia-NOAA** [comment]
It looks like the bias predictor variable names changed as part of [a recent UFO PR](https://github.com/JCSDA-internal/ufo/pull/3015). I resolved the `lapse_rate does not exist` error by updating the predictor names in the yaml as follows:

```
       obs bias:
         input file:  data/obs/atms_npp.satbias.nc4
         output file: ./out_atms_npp.satbias.nc
         variational bc:
           predictors:
           - name: constant
           - name: lapseRate
             order: 2
             tlapse: &atms_npp_tlapse  data/obs/atms_npp.tlapse.txt
           - name: lapseRate
             tlapse: *atms_npp_tlapse
           - name: emissivityJacobian
           - name: sensorScanAngle
             order: 4
           - name: sensorScanAngle
             order: 3
           - name: sensorScanAngle
             order: 2
           - name: sensorScanAngle
```

However, I am now getting a new error: 

```
 13: HDF5-DIAG: Error detected in HDF5 (1.14.0) MPI-process 13:
 13:   #000: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5D.c line 402 in H5Dopen2(): unable to synchronously open dataset
 13:     major: Dataset
 13:     minor: Can't open object
 13:   #001: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5D.c line 363 in H5D__open_api_common(): unable to open dataset
 13:     major: Dataset
 13:     minor: Can't open object
 13:   #002: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5VLcallback.c line 1980 in H5VL_dataset_open(): dataset open failed
 13:     major: Virtual Object Layer
 13:     minor: Can't open object
```

It looks like updating the submodules in #109 has broken some things in this yaml. I will continue to investigate.

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (SamuelDegelia-NOAA) [inline rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml]: - Change `lapse_rate` to `lapseRate`
- Change `emissivity` to `emissivityJacobian`
- Change `scan_angle` to `sensorScanAngle`

**Comment by xyzemc** [comment]
> It looks like the bias predictor variable names changed as part of [a recent UFO PR](https://github.com/JCSDA-internal/ufo/pull/3015). I resolved the `lapse_rate does not exist` error by updating the predictor names in the yaml as follows:
> 
> ```
>        obs bias:
>          input file:  data/obs/atms_npp.satbias.nc4
>          output file: ./out_atms_npp.satbias.nc
>          variational bc:
>            predictors:
>            - name: constant
>            - name: lapseRate
>              order: 2
>              tlapse: &atms_npp_tlapse  data/obs/atms_npp.tlapse.txt
>            - name: lapseRate
>              tlapse: *atms_npp_tlapse
>            - name: emissivityJacobian
>            - name: sensorScanAngle
>              order: 4
>            - name: sensorScanAngle
>              order: 3
>            - name: sensorScanAngle
>              order: 2
>            - name: sensorScanAngle
> ```
> 
> However, I am now getting a new error:
> 
> ```
>  13: HDF5-DIAG: Error detected in HDF5 (1.14.0) MPI-process 13:
>  13:   #000: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5D.c line 402 in H5Dopen2(): unable to synchronously open dataset
>  13:     major: Dataset
>  13:     minor: Can't open object
>  13:   #001: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5D.c line 363 in H5D__open_api_common(): unable to open dataset
>  13:     major: Dataset
>  13:     minor: Can't open object
>  13:   #002: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-hdf5-1.14.0-lixiejptsyy5mq3ocq6j2xx73mzd4xsl/spack-src/src/H5VLcallback.c line 1980 in H5VL_dataset_open(): dataset open failed
>  13:     major: Virtual Object Layer
>  13:     minor: Can't open object
> ```
> 
> It looks like updating the submodules in #109 has broken some things in this yaml. I will continue to investigate.

Thanks for digging into this.

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (xyzemc) [reply rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml]: > * Change `lapse_rate` to `lapseRate`
> * Change `emissivity` to `emissivityJacobian`
> * Change `scan_angle` to `sensorScanAngle`

resolved!

**Comment by SamuelDegelia-NOAA** [comment]
@xyzemc It looks like this error is also related to the VarBC changes (see [UFO #3015](https://github.com/JCSDA-internal/ufo/pull/3015) and [IODA #1092](https://github.com/JCSDA-internal/ioda/pull/1092)). JEDI now expects different variable names in the bias correction file (e.g., `BiasCoefficients` instead of `bias_coefficients`). Is it possible to regenerate the bias correction files using the latest versions of the submodules in RDASApp? I am not sure if the obs file needs to be recreated too.

**Comment by xyzemc** [comment]
> @xyzemc It looks like this error is also related to the VarBC changes (see [UFO #3015](https://github.com/JCSDA-internal/ufo/pull/3015) and [IODA #1092](https://github.com/JCSDA-internal/ioda/pull/1092)). JEDI now expects different variable names in the bias correction file (e.g., `BiasCoefficients` instead of `bias_coefficients`). Is it possible to regenerate the bias correction files using the latest versions of the submodules in RDASApp? I am not sure if the obs file needs to be recreated too.

I will regenerate the bias file with the new submodule.

**Comment by xyzemc** [comment]
> > @xyzemc It looks like this error is also related to the VarBC changes (see [UFO #3015](https://github.com/JCSDA-internal/ufo/pull/3015) and [IODA #1092](https://github.com/JCSDA-internal/ioda/pull/1092)). JEDI now expects different variable names in the bias correction file (e.g., `BiasCoefficients` instead of `bias_coefficients`). Is it possible to regenerate the bias correction files using the latest versions of the submodules in RDASApp? I am not sure if the obs file needs to be recreated too.
> 
> I will regenerate the bias file with the new submodule.

New satbias files are ready at: : /scratch1/NCEPDEV/fv3-cam/Xiaoyan.Zhang/mpas_jedi_conus/bufr2ioda/ioda_satbias_files
Please copy atms_npp.satbias.nc4 and atms_npp.satbias_cov.nc4

**Comment by guoqing-noaa** [comment]
> New satbias files are ready at: : /scratch1/NCEPDEV/fv3-cam/Xiaoyan.Zhang/mpas_jedi_conus/bufr2ioda/ioda_satbias_files Please copy atms_npp.satbias.nc4 and atms_npp.satbias_cov.nc4

Just curious, what does "cov" mean? "covariance"?

**Comment by xyzemc** [comment]
> > New satbias files are ready at: : /scratch1/NCEPDEV/fv3-cam/Xiaoyan.Zhang/mpas_jedi_conus/bufr2ioda/ioda_satbias_files Please copy atms_npp.satbias.nc4 and atms_npp.satbias_cov.nc4
> 
> Just curious, what does "cov" mean? "covariance"?

yes

**Comment by guoqing-noaa** [comment]
Thanks, @xyzemc!  Could you further help me understand why there is "covariance" related to satellite bias correction?

**Comment by xyzemc** [comment]
> Thanks, @xyzemc! Could you further help me understand why there is "covariance" related to satellite bias correction?

image

**Comment by guoqing-noaa** [comment]
Thank you! @xyzemc

**Comment by SamuelDegelia-NOAA** [comment]
Everything looks good to me now after @xyzemc modified the bias correction files for recent updates to VarBC in UFO/IODA. I was able to recreate @xyzemc's plot in the latest version of RDASApp (see below). . 

The new ATMS data are now staged in RDAS_DATA on Hera/Jet and will be synced to Orion/Hercules once those machines are back from maintenace. I think this PR is good to go if someone from GSL would like to have a look (@guoqing-noaa, @HaidaoLin-NOAA).

Increment plot:

**Review by SamuelDegelia-NOAA — DISMISSED** [review]

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (SamuelDegelia-NOAA) [inline rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml]: Oops, missed this one line. Could you change from `lapse_rate` to `lapseRate` @xyzemc?

**Comment by ShunLiu-NOAA** [comment]
Thank you @xyzemc and @SamuelDegelia-NOAA. @HaidaoLin-NOAA and @hu5970 please review this PR.

**Comment by xyzemc** [comment]
> Everything looks good to me now after @xyzemc modified the bias correction files for recent updates to VarBC in UFO/IODA. I was able to recreate @xyzemc's plot in the latest version of RDASApp (see below).
> 
> EDIT: See below for one more change that needs to be made. Once that is done I will approve the PR.
> 
> The new ATMS data are now staged in RDAS_DATA on Hera/Jet and will be synced to Orion/Hercules once those machines are back from maintenace. I think this PR is good to go if someone from GSL would like to have a look (@guoqing-noaa, @HaidaoLin-NOAA).
> 
> Increment plot: 

Thanks @SamuelDegelia-NOAA  for validate the result and everything else. I have committed the change of the last 'lapse_rate' to 'lapseRate' in the yaml file.

**Review by SamuelDegelia-NOAA — DISMISSED** [review]
Everything works for a fresh clone now. Thanks @xyzemc! 

Note: to test this, I replaced the obs-space section in `RDASApp/expr/mpas_2024052700/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml` with the yaml added in this PR. Then just ran the `run_jedi.sh` job in that directory.

**Comment by guoqing-noaa** [comment]
@xyzemc To confirm, do we need `atms_npp.satbias_cov.nc4` and `atms_npp.satbias.nc4` exist at the same time? Or we only need 'atms_npp.satbias_cov.nc4'

File: rrfs-test/CMakeLists.txt
Comment (guoqing-noaa) [inline rrfs-test/CMakeLists.txt]: @xyzemc could you remove this line since this PR does not change this CMakeLists.txt file?

File: rrfs-test/scripts/link_mpasjedi_expr.sh
Comment (guoqing-noaa) [inline rrfs-test/scripts/link_mpasjedi_expr.sh]: I suggest we don't use the version information in the run directory.
i.e: `ln -snf ${RDASApp}/fix/crtm/2.4.0 crtm_2.4.0`  will be good.
When we want to use a new CRTM version, we only need to update the link and don't need to update the yaml file.

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (guoqing-noaa) [inline rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml]: suggest changing `crtm_2.4.0` to `crtm` as proposed in the above comment.

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (guoqing-noaa) [inline rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml:581]: @SamuelDegelia-NOAA Do we need to keep the reference section here?

**Review by guoqing-noaa — CHANGES_REQUESTED** [review]
I made a few comments. Thanks!

**Comment by xyzemc** [comment]
> @xyzemc To confirm, do we need `atms_npp.satbias_cov.nc4` and `atms_npp.satbias.nc4` exist at the same time? Or we only need 'atms_npp.satbias_cov.nc4'

Yes, JEDI will read in both files at current version. It is possible change in the future version.

File: rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml
Comment (SamuelDegelia-NOAA) [reply rrfs-test/validated_yamls/templates/obtype_config/atms_npp_qc_bc.yaml:581]: No, we should not keep the reference section for individual obs space yamls like this. The plan (that I am currently working on testing) is to combine these individual obs space yamls into one larger yaml for each DA method (EnVar, GETKF, etc.). Then those larger yamls will become ctests and have their own reference section.

**Review by guoqing-noaa — DISMISSED** [review]

**Comment by xyzemc** [comment]
@guoqing-noaa can you recheck my latest change? I think I committed this after you approve this PR.

**Review by SamuelDegelia-NOAA — APPROVED** [review]

**Review by guoqing-noaa — APPROVED** [review]

**Review by ShunLiu-NOAA — APPROVED** [review]

**Pull request merged**
